src/functional/ffront/type_info.py

```diff
@@ -165,31 +166,123 @@ def is_concretizable(symbol_type: ct.SymbolType, to_type: ct.SymbolType) -> bool
+            raise NotImplementedError("Shape promotion not implemented.")
+        return types[0]
+    elif all(isinstance(type_, (ct.ScalarType, ct.FieldType)) for type_ in types):
+        dims = promote_dims(*(extract_dims(type_) for type_ in types))
+        dtype = cast(ct.ScalarType, promote(*(extract_dtype(type_) for type_ in types)))
+
+        return ct.FieldType(dims=dims, dtype=dtype)
+    raise TypeError("Expected a FieldType or ScalarType.")
+
+
+def promote_dims(
+    *dims_list: list[Dimension] | EllipsisType,
+) -> list[Dimension] | EllipsisType:
     """
-    Check if `symbol_type` has no fixed dimensionality and can be dimensionally promoted.
+    Find a unique ordering of multiple (individually ordered) lists of dimensions.
 
-    This is not to be mistaken for broadcasting, a more general concept, which allows interactions
-    between differing sets of dimensions.
+    The resulting list of dimensions contains all dimensions of the arguments
+    in the order they originally appear. If no unique order exists or a
+    contradicting order is found an exception is raised.
+
+    A modified version (ensuring uniqueness of the order) of
+    `Kahn's algirithm <https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm>`_
+    is used to topologically sort the arguments.
+
+    >>> I, J, K = (Dimension(value=dim) for dim in ["I", "J", "K"])
+    >>> promote_dims([I, J], [I, J, K]) == [I, J, K]
+    True
+    >>> promote_dims([I, J], [K]) # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+     ...
+    functional.common.GTTypeError: Dimensions can not be promoted. Could not determine order of the following dimensions: J, K.
+    >>> promote_dims([I, J], [J, I]) # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+     ...
+    functional.common.GTTypeError: Dimensions can not be promoted. The following dimensions appear in contradicting order: I, J.
     """
-    # scalars can always be broadcasted
-    if type_class(symbol_type) is ct.ScalarType:
-        return True
-    # symbol_type must be either zero- or any dimensional or have same dimensionality as to_type
-    elif type_class(to_type) is ct.FieldType:
-        dims = extract_dims(symbol_type)
-        return dims in [[], Ellipsis] or _is_sublist(dims, extract_dims(to_type))
-    return False
+    # build a graph with the vertices being dimensions and edges representing
+    #  the order between two dimensions
+    graph: dict[Dimension, set[Dimension]] = {}
+    for dims in dims_list:
+        if dims == Ellipsis:
+            return Ellipsis
+        dims = cast(list[Dimension], dims)
+        if len(dims) == 0:
+            continue
+        # create a vertex for each dimension
+        for dim in dims:
+            if dim not in graph:
+                graph[dim] = set()
+        # add edges
+        previous_dim = dims[0]
+        for dim in dims[1:]:
+            graph[dim].add(previous_dim)
+            previous_dim = dim
```

Comment:
Just a suggestion again, feel free to ignore:
```suggestion
        for d in range(1, len(dims)):
            graph[dim[d]].add(dim[d - 1])
```